db/types/operations/cast.py

```diff
@@ -255,6 +266,56 @@ def _escape_illegal_characters(sql_name):
         resulting_string = resulting_string.replace(old, new)
     return resulting_string
 
+def _get_json_type_body_map(target_type):
+    '''
+    json --> jsonb(or jsonb --> json): use default casting
+    text --> json: use _get_text_to_json_cast()
+    '''
+    default_behavior_source_types = frozenset([PostgresType.JSON, PostgresType.JSONB])
+    source_text_types = categories.STRING_TYPES
+    target_type_str = target_type.id
+
+    def _get_text_to_json_cast(target_type):
+        if target_type_str == "json":
+            return f"""
+                DECLARE integer_res {target_type_str};
+                BEGIN
+                    SELECT 
+                        CASE WHEN $1 IS NULL THEN NULL
+                        ELSE $1::{target_type_str} 
+                        END 
+                        INTO json_res;
+                IF json_typeof(json_res) != "object" AND  json_typeof(json_res) != "array"
+                    THEN RAISE EXCEPTION 'Invalid json expression';
+                END IF;
+                RETURN json_res;
+                END;  
+                """
+        else:
+            return f"""
+            DECLARE integer_res {target_type_str};
+                BEGIN
+                    SELECT 
+                        CASE WHEN $1 IS NULL THEN NULL
+                        ELSE $1::{target_type_str} 
+                        END 
+                        INTO json_res;
+                IF jsonb_typeof(json_res) != "object" AND  jsonb_typeof(json_res) != "array"
+                    THEN RAISE EXCEPTION 'Invalid json expression';
+                END IF;
+                RETURN json_res;
+                END;  
+                """
+
```

Comment:
1. I don't think if-else are necessary here
2. There's a bug, since you're declaring a variable called `integer_res`, but then assigning to a variable called `json_res`.